In [ ]:
!pip install catboost

### Importing required libraries

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import KFold, StratifiedKFold, ShuffleSplit, RepeatedStratifiedKFold

from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.tree import DecisionTreeClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from sklearn.neighbors import KNeighborsClassifier


In [4]:
df=pd.read_csv('/content/heart.csv')

In [5]:
df.head(5)

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

In [6]:
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

# Let's apply different model without taking care of categorical variable

In [7]:
X=df.drop(['target'],axis=1)
y=df['target']

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X,y,stratify=y)

In [9]:
svc=SVC()
svc.fit(X_train,y_train)
svc.score(X_test,y_test)

0.6052631578947368

In [10]:
lr=LogisticRegression()
lr.fit(X_train,y_train)
lr.score(X_test,y_test)

0.8552631578947368

In [11]:
nvb=GaussianNB()
nvb.fit(X_train,y_train)
nvb.score(X_test,y_test)

0.8157894736842105

In [12]:
catb=CatBoostClassifier()
catb.fit(X_train,y_train)
catb.score(X_test,y_test)

Learning rate set to 0.00547
0:	learn: 0.6882671	total: 49.4ms	remaining: 49.3s
1:	learn: 0.6845285	total: 50.2ms	remaining: 25.1s
2:	learn: 0.6807706	total: 51.1ms	remaining: 17s
3:	learn: 0.6767686	total: 51.9ms	remaining: 12.9s
4:	learn: 0.6730121	total: 52.7ms	remaining: 10.5s
5:	learn: 0.6689566	total: 53.6ms	remaining: 8.88s
6:	learn: 0.6652710	total: 54.4ms	remaining: 7.71s
7:	learn: 0.6608040	total: 55.2ms	remaining: 6.84s
8:	learn: 0.6568011	total: 56ms	remaining: 6.17s
9:	learn: 0.6522910	total: 56.9ms	remaining: 5.63s
10:	learn: 0.6477266	total: 57.7ms	remaining: 5.19s
11:	learn: 0.6443587	total: 58.5ms	remaining: 4.82s
12:	learn: 0.6412440	total: 59.3ms	remaining: 4.5s
13:	learn: 0.6368811	total: 60.1ms	remaining: 4.24s
14:	learn: 0.6326218	total: 61ms	remaining: 4s
15:	learn: 0.6284806	total: 61.8ms	remaining: 3.8s
16:	learn: 0.6240065	total: 62.6ms	remaining: 3.62s
17:	learn: 0.6208238	total: 63.5ms	remaining: 3.46s
18:	learn: 0.6173274	total: 64.3ms	remaining: 3.32s
19:	

0.8289473684210527

In [13]:
xgb=XGBClassifier()
xgb.fit(X_train,y_train)
xgb.score(X_test,y_test)

0.75

In [14]:
lgbm=LGBMClassifier()
lgbm.fit(X_train,y_train)
lgbm.score(X_test,y_test)

0.7894736842105263

In [15]:
model=KNeighborsClassifier(n_neighbors=1)
model.fit(X_train,y_train)
model.score(X_test,y_test)

#less score than others

0.5921052631578947

*Above models are around has descent score but we didn't tuned them.,   
In next series of code we'll be doing hyper-parameter tuning.*

## Let's improve our KNN model which previously gave 60% score

In [16]:
kfold=KFold(n_splits=5)
skfold=StratifiedKFold(n_splits=5,shuffle=True)
ss=ShuffleSplit(n_splits=8,train_size=.4,test_size=.3)
rs=RepeatedStratifiedKFold(n_splits=5,n_repeats=10)

In [17]:
print("KFold: ",cross_val_score(KNeighborsClassifier(),X,y,cv=kfold))
print("skfold: ",cross_val_score(KNeighborsClassifier(),X,y,cv=skfold))
print("ShuffleSplit: ",cross_val_score(KNeighborsClassifier(),X,y,cv=ss))
print("RepeatedStratifiedKFold: ",cross_val_score(KNeighborsClassifier(),X,y,cv=rs))


KFold:  [0.47540984 0.63934426 0.67213115 0.38333333 0.31666667]
skfold:  [0.57377049 0.68852459 0.73770492 0.68333333 0.65      ]
ShuffleSplit:  [0.6043956  0.65934066 0.68131868 0.59340659 0.53846154 0.59340659
 0.63736264 0.65934066]
RepeatedStratifiedKFold:  [0.68852459 0.59016393 0.73770492 0.68333333 0.63333333 0.75409836
 0.63934426 0.68852459 0.68333333 0.58333333 0.59016393 0.72131148
 0.59016393 0.68333333 0.63333333 0.63934426 0.62295082 0.63934426
 0.61666667 0.66666667 0.60655738 0.55737705 0.68852459 0.68333333
 0.65       0.50819672 0.63934426 0.67213115 0.66666667 0.68333333
 0.70491803 0.55737705 0.54098361 0.66666667 0.66666667 0.67213115
 0.60655738 0.62295082 0.75       0.66666667 0.72131148 0.62295082
 0.63934426 0.58333333 0.6        0.60655738 0.55737705 0.72131148
 0.68333333 0.58333333]


*  *Best KNN score is nearly 75% in RepeatedStratifiedKFold*

In [18]:
from sklearn.model_selection import GridSearchCV
param_grid={'max_depth':range(1,7)}
grid=GridSearchCV(DecisionTreeClassifier(random_state=0),param_grid=param_grid,cv=10)
grid.fit(X_train,y_train)

GridSearchCV(cv=10, error_score=nan,
             estimator=DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort='deprecated',
                                              random_state=0, splitter='best'),
             iid='deprecated', n_jobs=None,
             param_grid={'max_depth': range(1, 7)}, pre_dispatch='2*n_jobs',
             refit=True, return_train_score=False

In [19]:
print("Best Score: ",grid.score(X_test,y_test))

Best Score:  0.7763157894736842


*  *Even decision tree gave us best score of 77, Let's improve our KNN model more*

# We know that KNN is distance based matric and so this time we are going to do some data preprocessing

In [20]:
data=pd.get_dummies(df,columns=['cp','fbs','restecg','exang','slope','ca','thal'])
data['trestbps']=data['trestbps']/200
data['chol']=data['chol']/564
data['thalach']=data['thalach']/564

In [21]:
data.head(5)

age  sex  trestbps      chol   thalach  ...  ca_4  thal_0  thal_1  thal_2  thal_3
0   63    1     0.725  0.413121  0.265957  ...     0       0       1       0       0
1   37    1     0.650  0.443262  0.331560  ...     0       0       0       1       0
2   41    0     0.650  0.361702  0.304965  ...     0       0       0       1       0
3   56    1     0.600  0.418440  0.315603  ...     0       0       0       1       0
4   57    0     0.600  0.627660  0.289007  ...     0       0       0       1       0

[5 rows x 30 columns]

In [22]:
df.head(5)

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

In [23]:
data2=data.drop(['cp_0','fbs_0','restecg_0','exang_0','slope_0','ca_0','thal_0'],axis=1)

In [24]:
X=data2.drop(['target','age'],axis=1)
y=data2['target']
kfold=KFold(n_splits=5)
skfold=StratifiedKFold(n_splits=5,shuffle=True)
ss=ShuffleSplit(n_splits=8,train_size=.4,test_size=.3)
rs=RepeatedStratifiedKFold(n_splits=5,n_repeats=10)

In [25]:
print("KFold: ",cross_val_score(KNeighborsClassifier(),X,y,cv=kfold))
print("skfold: ",cross_val_score(KNeighborsClassifier(),X,y,cv=skfold))
print("ShuffleSplit: ",cross_val_score(KNeighborsClassifier(),X,y,cv=ss))
print("RepeatedStratifiedKFold: ",cross_val_score(KNeighborsClassifier(),X,y,cv=rs))

KFold:  [0.78688525 0.70491803 0.85245902 0.7        0.58333333]
skfold:  [0.80327869 0.7704918  0.86885246 0.78333333 0.78333333]
ShuffleSplit:  [0.83516484 0.85714286 0.82417582 0.76923077 0.78021978 0.84615385
 0.9010989  0.72527473]
RepeatedStratifiedKFold:  [0.75409836 0.85245902 0.86885246 0.8        0.81666667 0.7704918
 0.78688525 0.90163934 0.81666667 0.8        0.75409836 0.80327869
 0.78688525 0.83333333 0.78333333 0.83606557 0.78688525 0.78688525
 0.83333333 0.76666667 0.86885246 0.81967213 0.81967213 0.8
 0.78333333 0.91803279 0.90163934 0.83606557 0.61666667 0.81666667
 0.7704918  0.85245902 0.78688525 0.81666667 0.86666667 0.7704918
 0.8852459  0.81967213 0.85       0.8        0.78688525 0.83606557
 0.7704918  0.8        0.75       0.78688525 0.78688525 0.81967213
 0.83333333 0.81666667]


In [26]:
cross_val_scores=[]
X_train, X_test, y_train, y_test = train_test_split(X,y,stratify=y)

for i in range(1,15,3):
  model=KNeighborsClassifier(n_neighbors=i)
  scores=cross_val_score(model, X_train,y_train,cv=10)
  cross_val_scores.append(np.mean(scores))

In [27]:
print("Best Score: ",np.max(cross_val_scores))

Best Score:  0.8100790513833992


In [28]:
X_train, X_test, y_train, y_test = train_test_split(X,y,stratify=y)

param_grid={'n_neighbors':np.arange(1,25,3)}

gridSearch=GridSearchCV(KNeighborsClassifier(),param_grid,cv=10,return_train_score=True)

gridSearch.fit(X_train,y_train)

print("Best Score TRAIN: ",gridSearch.best_score_)
print("Best Param: ",gridSearch.best_params_)
print("Best Score Test: ",gridSearch.score(X_test,y_test))

Best Score TRAIN:  0.8373517786561264
Best Param:  {'n_neighbors': 19}
Best Score Test:  0.8026315789473685


*  *Little bit of uncertain using such n_neighbors., Let's try out something more robust to figure out correct parameter*

In [29]:
from sklearn.metrics import accuracy_score

In [30]:
param_grid = [{'n_neighbors': np.arange(1, 15, 1)}]

In [31]:
kf = StratifiedKFold(n_splits=10,random_state=1,shuffle=True)
i=1
for train_index,test_index in kf.split(X,y):
     print('\n{} of kfold {}'.format(i,kf.n_splits))
     xtr,xvl = X.loc[train_index],X.loc[test_index]
     ytr,yvl = y[train_index],y[test_index]
     model = GridSearchCV(KNeighborsClassifier(), param_grid, cv=10, scoring= 'f1',iid=True)
     model.fit(xtr, ytr)
     print (model.best_params_)
     pred=model.predict(xvl)
     print('accuracy_score',accuracy_score(yvl,pred))
     i+=1


1 of kfold 10
{'n_neighbors': 7}
accuracy_score 0.8387096774193549

2 of kfold 10
{'n_neighbors': 5}
accuracy_score 0.7096774193548387

3 of kfold 10
{'n_neighbors': 13}
accuracy_score 0.7741935483870968

4 of kfold 10
{'n_neighbors': 13}
accuracy_score 0.7333333333333333

5 of kfold 10
{'n_neighbors': 7}
accuracy_score 0.8

6 of kfold 10
{'n_neighbors': 7}
accuracy_score 0.8666666666666667

7 of kfold 10
{'n_neighbors': 11}
accuracy_score 0.7666666666666667

8 of kfold 10
{'n_neighbors': 6}
accuracy_score 0.7666666666666667

9 of kfold 10
{'n_neighbors': 7}
accuracy_score 0.8666666666666667

10 of kfold 10
{'n_neighbors': 7}
accuracy_score 0.7666666666666667


*  *We got best score using n_neighbours=7 for most of the time*

#### So we got upto 86.6% of accuracy using proper parameter i.e. n_neighbors=7


### Let's try out more robust model "XGBClassifier"

In [32]:
param_grid = [{'min_child_weight': np.arange(0.1, 10.1, 0.1)}]

In [33]:
kf = StratifiedKFold(n_splits=10,random_state=1,shuffle=True)
i=1
for train_index,test_index in kf.split(X,y):
     print('\n{} of kfold {}'.format(i,kf.n_splits))
     xtr,xvl = X.loc[train_index],X.loc[test_index]
     ytr,yvl = y[train_index],y[test_index]
     model2 = GridSearchCV(XGBClassifier(), param_grid, cv=10, scoring= 'f1',iid=True)
     model2.fit(xtr, ytr)
     print (model2.best_params_)
     pred=model2.predict(xvl)
     print('accuracy_score',accuracy_score(yvl,pred))
     i+=1


1 of kfold 10
{'min_child_weight': 3.8000000000000003}
accuracy_score 0.8709677419354839

2 of kfold 10
{'min_child_weight': 5.0}
accuracy_score 0.7419354838709677

3 of kfold 10
{'min_child_weight': 4.6}
accuracy_score 0.7741935483870968

4 of kfold 10
{'min_child_weight': 1.7000000000000002}
accuracy_score 0.8

5 of kfold 10
{'min_child_weight': 1.6}
accuracy_score 0.9

6 of kfold 10
{'min_child_weight': 1.4000000000000001}
accuracy_score 0.8333333333333334

7 of kfold 10
{'min_child_weight': 4.5}
accuracy_score 0.8

8 of kfold 10
{'min_child_weight': 4.1}
accuracy_score 0.7333333333333333

9 of kfold 10
{'min_child_weight': 0.9}
accuracy_score 0.8666666666666667

10 of kfold 10
{'min_child_weight': 3.2}
accuracy_score 0.7666666666666667


#### We got around 83% to 90% accuracy using proper parameter for min_child_weight=1.6